microindex/writer.go

```diff
@@ -78,45 +79,76 @@ type indexWriter struct {
 // provide keys in increasing lexicographic order.  Duplicate keys are not
 // allowed but will not be detected.  Close() or Abort() must be called when
 // done writing.
-func NewWriter(zctx *resolver.Context, path string, keyFields []field.Static, frameThresh int) (*Writer, error) {
-	return NewWriterWithContext(context.Background(), zctx, path, keyFields, frameThresh)
-
+func NewWriter(zctx *resolver.Context, path string, options ...Option) (*Writer, error) {
+	return NewWriterWithContext(context.Background(), zctx, path, options...)
 }
-func NewWriterWithContext(ctx context.Context, zctx *resolver.Context, path string, keyFields []field.Static, frameThresh int) (*Writer, error) {
-	if keyFields == nil {
-		keyFields = []field.Static{field.New("key")}
+
+func NewWriterWithContext(ctx context.Context, zctx *resolver.Context, path string, options ...Option) (*Writer, error) {
+	w := &Writer{zctx: zctx}
+	for _, opt := range options {
+		opt.apply(w)
+	}
+	if w.keyFields == nil {
+		w.keyFields = []field.Static{field.New("key")}
 	}
-	if frameThresh == 0 {
-		return nil, errors.New("microindex frame threshold cannot be zero")
+	if w.frameThresh == 0 {
+		w.frameThresh = frameThresh
 	}
-	if frameThresh > FrameMaxSize {
-		return nil, fmt.Errorf("frame threshold too large (%d)", frameThresh)
+	if w.frameThresh > FrameMaxSize {
+		return nil, fmt.Errorf("frame threshold too large (%d)", w.frameThresh)
 	}
-	uri, err := iosrc.ParseURI(path)
+	var err error
+	w.uri, err = iosrc.ParseURI(path)
 	if err != nil {
 		return nil, err
 	}
-	writer, err := iosrc.NewWriter(ctx, uri)
+	w.iow, err = iosrc.NewWriter(ctx, w.uri)
 	if err != nil {
 		return nil, err
 	}
-	tmp, err := ioutil.TempDir("", "microindex-*")
+	w.tmpdir, err = ioutil.TempDir("", "microindex-*")
 	if err != nil {
 		return nil, err
 	}
-	fields, resolvers := expr.CompileAssignments(keyFields, keyFields)
-	w := &Writer{
-		zctx:        zctx,
-		uri:         uri,
-		cutter:      cut.NewStrictCutter(zctx, false, fields, resolvers),
-		tmpdir:      tmp,
-		keyFields:   keyFields,
-		frameThresh: frameThresh,
-		iow:         writer,
-	}
+	fields, resolvers := expr.CompileAssignments(w.keyFields, w.keyFields)
+	w.cutter = cut.NewStrictCutter(zctx, false, fields, resolvers)
 	return w, nil
 }
 
+type Option interface {
+	apply(*Writer)
+}
+
+type optionFunc func(*Writer)
+
+func (f optionFunc) apply(w *Writer) { f(w) }
```

Comment:
Any reason not to just say `type Option func(*Writer)` and do away with optionFunc?

Reply:
In the grand scheme of things probably no. When  I look at it through golang docs I like the way it looks better. Also I like that it obscures the internals, so consumers of it don't try to make their own options. That said, the stakes are low.